docs/cudf/source/developer_guide/ARCHITECTURE.md

```diff
@@ -0,0 +1,199 @@
+Most `GenericIndex` methods are inherited from `Frame`, saving us the trouble of rewriting them.
+
+We now consider the three main exceptions to this model:
+
+- A `RangeIndex` is not backed by a column of data, so it inherits directly from `BaseIndex` alone.
+  Wherever possible, its methods have special implementations designed to avoid materializing columns.
+  Where such an implementation is infeasible, we fall back to converting it to an integer index first instead.
+- A `MultiIndex` is backed by _multiple_ columns of data.
+  Therefore, its inheritance hierarchy looks like `Frame`->``MultiIndex`<-`BaseIndex`.
+  Some of its more `Frame`-like methods may be inherited,
+  but many others must be reimplemented since in many cases a `MultiIndex` is not expected to behave like a `Frame`.
+- Just like in pandas, `Index` itself can never be instantiated.
+  `pandas.Index` is the parent class for indexes,
+  but its constructor returns an appropriate subclass depending on the input data type and shape.
+  Unfortunately, mimicking this behavior requires overriding `__new__`,
+  which in turn makes shared intialization across inheritance trees much more cumbersome to manage.
+  To reenable sharing constructor logic across different index classes,
+  we instead define `BaseIndex` as the parent class of all indexes.
+  `Index` inherits from `BaseIndex`, but it masquerades as a `BaseIndex` to match pandas.
+  This class should contain no implementations since it is simply a factory for other indexes.
+
+
+## The Column layer
+
+The next layer in the cuDF stack is the Column layer.
+This layer forms the glue between pandas-like APIs and our underlying data layouts.
+Under the hood, cuDF is built around the [Apache Arrow Format](https://arrow.apache.org).
+This data format is both conducive to high-performance algorithms and suitable for data interchange between libraries.
+
+The principal objects in the Column layer are the `ColumnAccessor` and the various `Column` classes.
+The `Column` is cuDF's core data structure and is modeled after the
+[Apache Arrow Columnar Format](https://arrow.apache.org/docs/format/Columnar.html).
+A `ColumnAccessor` is a dictionary-like interface to a sequence of `Columns`.
+A `Frame` owns a `ColumnAccessor`, and most of its operations are implemented as loops over that object's `Column`s.
+
+### ColumnAccessor
+
+The primary purpose of the `ColumnAccessor` is to encapsulate pandas column selection semantics.
+Columns may be selected or inserted by index or by label, and label-based selections are as flexible as pandas is.
+For instance, Columns may be selected hierarchically (using tuples) or via wildcards.
+`ColumnAccessors` also support the `MultiIndex` columns that can result from operations like groupbys.
+
+### Columns
+
+The parent `Column` class is implemented in Cython to support interchange with C++ (more on that later).
+However, the bulk of the `Column`'s functionality is embodied by the `ColumnBase` subclass.
```

Comment:
Is it weird that `ColumnBase` derives from `Column`, while `Index` derives from `BaseIndex`? Two reasons:
- The "base" isn't actually the base class for columns.
- The word "Base" appears in different places in each class name (beginning / end). Should this be `BaseColumn`?

Reply:
It is weird, but it is correct. The two evolved completely independently. The reasons for the naming choices are esoteric and not really worth discussing in this document. For now, I think we just live with the names being what they are.

Reply:
Yes -- I agree the docs are aligned with the implementation. Should we file an issue to mention this and shoot for a longer-term refactoring towards consistent naming?

Reply:
I tried looking in the CPython codebase for `Base` in a class name and came up with mixed results.

Reply:
We can file an issue. My hope is that the long term resolution of this will instead involve removing the current inheritance and not having separate `Column` and `ColumnBase` classes.

Reply:
See #10869.